### ISSUE TITLE
Refactor PouchDB usage - use a CLS function context instead

### ISSUE THREAD

**Issue by mike12345567** [Budibase/budibase#4195]
## Description
Been thinking lately about our usage of `new CouchDB(...)` and whether or not it could be improved.
The TLDR of this issue is to shift to a mechanism like we use for tenancy, with the `cls-hooked` function context for the app ID and the app database. The idea of this is to use a `getDevAppDB()` or `getProdAppDB()` call, as in the server 99% of the database access is to either the prod app, dev app or global databases. There should never be cross app communicate either, so this should be restrictive in a good way.

There are two key benefits to doing this:
1. Code quality - right now there is a bit of an anti-pattern that exists around some of the very expansive APIs, where we'll sometimes pass the db instance around and sometimes we'll pass the appId around - which can be painful if you are trying to integrate parts of the API and they don't do it the same way, you have to go refactor one to follow a single pattern so that they can talk. If we moved to a method where a single contextual method is used to retrieve the DB everyone would be able to speak the same language easily.
2. Performance - it turns out there is a negative to our current approach, everytime a new instance is created of Pouch it carries out an `info()` call to see if the database exists or not. For very heavy endpoints there could be a large number of excess HTTP calls, which serve no value.

Specific examples of APIs that would be affected by this (performance wise) are things like say the link controller, view functions, or bulk operations called by table ops, where it'll pass the app ID down to the functions to construct a new DB accessor.

All in all I think this could be worthwhile, we could expose three things that we use commonly within the context:
1. The app ID, so that it doesn't need to be passed around everywhere, simplifying function signatures.
2. The prod app DB
3. The dev app DB

I think the way the DBs should work is that the context functions should check if it exists in the context when the `getProdAppDB` or `getDevAppDB` functions are called, if they don't create them and add to context, that way per API call there will be max 2 DB instances created (a prod and a dev one - e.g. for replication).

## Investigation
I did some checks about the performance implications of this to check if its true or not (the performance issue) - when I wiresharked the connection between the database and the server I noticed numerous calls like:
```json
HTTP/1.1 200 OK
Cache-Control: must-revalidate
Connection: close
Content-Length: 481
Content-Type: application/json
Date: Tue, 25 Jan 2022 15:26:23 GMT
Server: CouchDB/3.1.1 (Erlang OTP/20)
X-Couch-Request-ID: aec96cf81c
X-CouchDB-Body-Time: 0
{"db_name":"app_dev_7aba821695584071a339975374daec6c","purge_seq":"0-g1AAAABXeJzLYWBgYMpgTmEQTM4vTc5ISXIwNDLXMwBCwxyQVB4LkGRoAFL_gSArkQGP2kSGpHqIoiwAtOgYRA","update_seq":"559-g1AAAABaeJzLYWBgYMpgTmEQTM4vTc5ISXIwNDLXMwBCwxyQVB4LkGRoAFL_gSArcSoetYkMSfUgRUkMDIyzsgAWNBl1","sizes":{"file":950702,"external":18906,"active":89670},"props":{},"doc_del_count":18,"doc_count":25,"disk_format_version":8,"compact_running":false,"cluster":{"q":2,"n":1,"w":1,"r":1},"instance_start_time":"0"}
```

I was careful to breakpoint the server and clear the logs, making sure the calls I was looking at where only for a few different `new CouchDB(appId)` calls.
